runtime/drive_handler.go

```diff
@@ -15,125 +15,194 @@ package main
-		if err := opt(drives); err != nil {
-			h.logger.WithError(err).Debug("failed to apply option to stub drives")
-			return nil, err
-		}
-	}
-	h.drives = drives
-	return &h, nil
+	freeDrive := h.freeDrives[0]
+	h.freeDrives = h.freeDrives[1:]
+	h.usedDrives[id] = freeDrive
+
+	return freeDrive.withMountConfig(
+		hostPath,
+		vmPath,
+		filesystemType,
+		options,
+	), nil
 }
 
-func (h *stubDriveHandler) createStubDrives(stubDriveCount int) ([]models.Drive, error) {
-	paths, err := h.stubDrivePaths(stubDriveCount)
-	if err != nil {
-		return nil, err
-	}
+// CreateDriveMountStubs creates a set of MountableStubDrives from the provided DriveMount configs.
+func CreateDriveMountStubs(
+	machineCfg *firecracker.Config,
+	jail jailer,
+	driveMounts []*proto.FirecrackerDriveMount,
+	logger *logrus.Entry,
+) ([]MountableStubDrive, error) {
+	containerStubs := make([]MountableStubDrive, len(driveMounts))
+	for i, driveMount := range driveMounts {
+		isReadOnly := driveMount.IsReadOnly
+		rateLimiter := driveMount.RateLimiter
+		stubFileName := fmt.Sprintf("drivemntstub%d", i)
+
+		stubDrive, err := newStubDrive(
+			filepath.Join(jail.JailPath().RootPath(), stubFileName),
+			jail, isReadOnly, rateLimiter, logger)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create drive mount stub drive")
+		}
 
-	stubDrives := make([]models.Drive, 0, stubDriveCount)
-	for i, path := range paths {
-		stubDrives = append(stubDrives, models.Drive{
-			DriveID:      firecracker.String(fmt.Sprintf("stub%d", i)),
-			IsReadOnly:   firecracker.Bool(false),
-			PathOnHost:   firecracker.String(path),
+		machineCfg.Drives = append(machineCfg.Drives, models.Drive{
+			DriveID:      firecracker.String(stubDrive.driveID),
+			PathOnHost:   firecracker.String(stubDrive.stubPath),
+			IsReadOnly:   firecracker.Bool(isReadOnly),
+			RateLimiter:  rateLimiterFromProto(rateLimiter),
 			IsRootDevice: firecracker.Bool(false),
 		})
+		containerStubs[i] = stubDrive.withMountConfig(
+			driveMount.HostPath,
+			driveMount.VMPath,
+			driveMount.FilesystemType,
+			driveMount.Options)
 	}
 
-	return stubDrives, nil
+	return containerStubs, nil
 }
 
-// stubDrivePaths will create stub drives and return the paths associated with
-// the stub drives.
-func (h *stubDriveHandler) stubDrivePaths(count int) ([]string, error) {
-	paths := []string{}
-	for i := 0; i < count; i++ {
-		driveID := fmt.Sprintf("stub%d", i)
-		path := filepath.Join(h.RootPath, driveID)
-
-		if err := h.createStubDrive(driveID, path); err != nil {
-			return nil, err
-		}
-
-		paths = append(paths, path)
-	}
+// A MountableStubDrive represents a stub drive that is ready to be patched and mounted
+// once PatchAndMount is called.
+type MountableStubDrive interface {
+	PatchAndMount(
+		requestCtx context.Context,
+		machine firecracker.MachineIface,
+		driveMounter drivemount.DriveMounterService,
+	) error
+}
 
-	return paths, nil
+func stubPathToDriveID(stubPath string) string {
```

Comment:
Can you explain which characters are problematic in a comment.

Reply:
Done

Reply:
Base64 will remove "/" but may contain "-" in the output. We could use Base32 instead.

Reply:
Good catch! I didn't know that. Switched to Base32 and also switched to using just the base name of the stub file (which is unique per shim/vm anyways) so the length of the drive ids doesn't get too outrageous.

Reply:
Can we add a test that has an invalid `stubPath` just to ensure that this doesn't regress?

Reply:
> Can we add a test that has an invalid stubPath just to ensure that this doesn't regress and is being used properly?

I'm not sure I follow how we could create an invalid stubPath as the implementation of our code now shouldn't have any invalid values for a stubPath. Would the test create a drive id with an invalid character and then provide that directly to the firecracker API? That wouldn't really be a test of our code.

The only thing I can think of here would be to do some kind of general fuzz testing on our api to try to find inputs that create errors, but that seems like a much, much larger effort outside of scope here.

Reply:
> Would the test create a drive id with an invalid character and then provide that directly to the firecracker API? That wouldn't really be a test of our code.

Yea, I was thinking this, just to ensure that it is encoding it properly. You should then be able to read the stub drive and compare the base32 value?

Reply:
Ohh you just mean this function in isolation, I see, I was thinking about some way to get an invalid stub path passed to this function. Yeah I will add a test that does a regex check on the output of this function for some paths with invalid chars.

Reply:
Updated with that test